tests/functional/tests/engine/test_pp.py

```diff
@@ -0,0 +1,338 @@
+    completions = []
+    for i in range(ios_to_issue):
+        comp = OcfCompletion([("error", c_int)])
+        write_data = Data(max_io_size)
+        io = core.new_io(
+            cache.get_default_queue(),
+            i * max_io_size,
+            write_data.size,
+            IoDir.WRITE,
+            0,
+            0,
+        )
+        io.set_data(write_data)
+        io.callback = comp.callback
+        completions += [comp]
+        io.submit()
+
+    if bytes_to_fill % max_io_size:
+        comp = OcfCompletion([("error", c_int)])
+        write_data = Data(
+            Size.from_B(bytes_to_fill % max_io_size, sector_aligned=True)
+        )
+        io = core.new_io(
+            cache.get_default_queue(),
+            ios_to_issue * max_io_size,
+            write_data.size,
+            IoDir.WRITE,
+            0,
+            0,
+        )
+        io.set_data(write_data)
+        io.callback = comp.callback
+        completions += [comp]
+        io.submit()
+
+    for c in completions:
+        c.wait()
+
+
+@pytest.mark.parametrize("fill_percentage", [0, 1, 50, 99])
+@pytest.mark.parametrize("insertion_threshold", [2, 8])
+@pytest.mark.parametrize("io_dir", IoDir)
+def test_promoted_after_hits_various_thresholds(
+    pyocf_ctx, io_dir, insertion_threshold, fill_percentage
+):
+    """
+    Check promotion policy behavior with various set thresholds
+
+    1. Create core/cache pair with promotion policy NHIT
+    2. Set TRIGGER_THRESHOLD/INSERTION_THRESHOLD to predefined values
+    3. Fill cache from the to TRIGGER_THRESHOLD
+        * occupancy percentage should equal to TRIGGER_THRESHOLD (all requests promoted)
+    4. Issue INSERTION_THRESHOLD - 1 requests to core line not inserted to cache
+        * occupancy should not change
+    5. Issue one request to LBA from step 4
+        * occupancy should rise by one cache line
+    """
+
+    # Step 1
+    cache_device = Volume(Size.from_MiB(30))
+    core_device = Volume(Size.from_MiB(30))
+
+    cache = Cache.start_on_device(
+        cache_device, promotion_policy=PromotionPolicy.NHIT
+    )
+    core = Core.using_device(core_device)
+    cache.add_core(core)
+
+    # Step 2
+    cache.set_promotion_policy_param(
+        NhitParams.TRIGGER_THRESHOLD, fill_percentage
+    )
+    cache.set_promotion_policy_param(
+        NhitParams.INSERTION_THRESHOLD, insertion_threshold
+    )
+    # Step 3
+    fill_cache(cache, fill_percentage / 100)
+
+    stats = cache.get_stats()
+    cache_lines = stats["conf"]["size"]
+    assert stats["usage"]["occupancy"]["permil"] == fill_percentage * 10
+    filled_occupancy = stats["usage"]["occupancy"]["value"]
+
+    # Step 4
+    completions = []
+    for i in range(insertion_threshold - 1):
+        comp = OcfCompletion([("error", c_int)])
+        write_data = Data(cache_lines.line_size)
+        io = core.new_io(
+            cache.get_default_queue(),
+            int(cache_lines.bytes) - cache_lines.line_size,
```

Comment:
Can you assign this to a variable with meaningful name? It's not immediately clear what this is.

Reply:
Done